logstash-core/src/main/java/org/logstash/ackedqueue/AckedReadBatch.java

```diff
@@ -21,89 +21,75 @@
 package org.logstash.ackedqueue;
 
 import org.jruby.RubyArray;
-import org.jruby.RubyHash;
 import org.logstash.ackedqueue.ext.JRubyAckedQueueExt;
 import org.logstash.execution.MemoryReadBatch;
 import org.logstash.execution.QueueBatch;
 import org.logstash.ext.JrubyEventExtLibrary.RubyEvent;
-
 import java.io.IOException;
 import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Set;
 
 import static org.logstash.RubyUtil.RUBY;
 
 public final class AckedReadBatch implements QueueBatch {
 
     private AckedBatch ackedBatch;
 
-    private RubyHash originals;
-
-    private RubyHash generated;
+    private Set<RubyEvent> events;
 
     public static AckedReadBatch create(
         final JRubyAckedQueueExt queue,
         final int size,
         final long timeout)
     {
-        return new AckedReadBatch(queue, size, timeout);
-    }
-
-    private AckedReadBatch(
-        final JRubyAckedQueueExt queue,
-        final int size,
-        final long timeout)
-    {
-        AckedBatch batch;
         try {
-            batch = queue.readBatch(size, timeout);
+            final AckedBatch batch = queue.readBatch(size, timeout);
+            return (batch == null) ? new AckedReadBatch() : new AckedReadBatch(batch);
         } catch (IOException e) {
             throw new IllegalStateException(e);
         }
-        if (batch == null) {
-            originals = RubyHash.newHash(RUBY);
-            ackedBatch = null;
-        } else {
-            ackedBatch = batch;
-            originals = ackedBatch.toRubyHash(RUBY);
-        }
-        generated = RubyHash.newHash(RUBY);
+    }
+
+    public static AckedReadBatch create() {
+        return new AckedReadBatch();
+    }
+
+    private AckedReadBatch() {
+        ackedBatch = null;
+        events = new LinkedHashSet<>();
```

Comment:
The events collection is used by the downstream DataSet implementations,  if we have no event I think that those DataSet doesn't change the content of the events collections, so it should be immutable, to enforce that could we use the Collections.emptySet() (or emptyList in case the other comments is valid)?

Reply:
Good catch.

Reply:
So, I don't think I'll change that, for 2 reasons:
- an `AckedReadBatch` is mutable through the `merge` method, so creating an immutable empty set would break that contract. Instead maybe we could create a `new LinkedHashSet(0)` to minimize allocations but still be able to call merge which is rare.
- These empty batches are essentially created when exiting the worker loop do perform a final flush so this would be a very low impact optimization. 
WDYT?

Reply:
I decided to refactor that, see comments.